src/Toolkit/Toolkit/UI/Controls/PopupViewer/NavigationSubView.cs

```diff
@@ -0,0 +1,284 @@
+                cc2.Content = content;
+            UpdateView();
+        }
+
+        private void GoBack()
+        {
+            if (NavigationStack.Count == 0)
+                return;
+            var previousPage = NavigationStack.Pop();
+            var content = previousPage.Item1;
+            var lastOffset = previousPage.Item2;
+
+#if WINDOWS_XAML
+            ContentTransitions = new TransitionCollection
+            {
+                new EntranceThemeTransition() { FromHorizontalOffset = -200, FromVerticalOffset = 0 }
+            };
+#endif
+            if (GetTemplateChild("ScrollViewer") is ScrollViewer sv)
+            {
+#if WINUI
+                // Restore scroll offset on back navigation
+                EventHandler<object>? handler = null;
+                handler = (s, e) =>
+                {
+                    sv.LayoutUpdated -= handler;
+                    sv.ChangeView(null, lastOffset, null, true);
+                };
+                sv.LayoutUpdated += handler;
+#elif WPF
+                ScrollChangedEventHandler? handler = null;
+                handler = (s, e) =>
+                {
+                    if (e.ExtentHeight >= lastOffset)
+                    {
+                        sv.ScrollChanged -= handler;
+                        sv.ScrollToVerticalOffset(lastOffset);
+                    }
+                };
+                sv.ScrollChanged += handler; 
+#elif MAUI
+                EventHandler? handler = null;
+                handler = (s, e) =>
+                {
+                    sv.Content.SizeChanged -= handler;
+                    _ = sv.ScrollToAsync(0, lastOffset, false);
+                };
+                sv.Content.SizeChanged += handler;
+#endif
+            }
+            SetContent(content);
+        }
+
+        private void GoUp()
+        {
+            if (NavigationStack.Count == 0)
+                return;
+            var content = NavigationStack.First();
+            NavigationStack.Clear();
+#if WINDOWS_XAML
+            ContentTransitions = new TransitionCollection
+            {
+                new EntranceThemeTransition() { FromHorizontalOffset = -200, FromVerticalOffset = 0 }
+            };
+#endif
+            SetContent(content);
+        }
+
+        /// <inheritdoc />
+#if WINDOWS_XAML || MAUI
+        protected override void OnApplyTemplate()
+#elif WPF
+        public override void OnApplyTemplate()
+#endif
+        {
+            base.OnApplyTemplate();
+            if (GetTemplateChild("NavigateBack") is Button backButton)
+            {
+#if MAUI
+                backButton.Clicked += (s, e) => GoBack();
+#else
+                backButton.Click += (s, e) => GoBack();
+#endif
+            }
+            if (GetTemplateChild("NavigateUp") is Button upButton)
+            {
+#if MAUI
+                upButton.Clicked += (s, e) => GoUp();
+#else
+                upButton.Click += (s, e) => GoUp();
+#endif
+            }
```

Comment:
Can template change more than once? If so, do we need to detach listeners from the old controls or use weak listeners?

Reply:
The event handler will go out of scope with the button, so no need here.